lib/pos_mempool_persister.go

```diff
@@ -121,32 +127,54 @@ func (mp *MempoolPersister) run() {
 // Stop is used to stop the persister thread and reset the persister state. It will wait for the persister thread to
 // flush the outstanding updateBatch to the database before returning.
 func (mp *MempoolPersister) Stop() error {
-	if mp.status == MempoolPersisterStatusNotRunning {
+	// stopLock ensures that only a single MempolEventExit is enqueued in case someone called Stop concurrently.
+	mp.stopLock.Lock()
```

Comment:
Why not just hold `mp.Lock()` here?

Reply:
Ah this one is tricky. I don't like this second lock either. But I couldn't find an alternative that is thread safe on concurrent `Stop()` calls. 

The issue is that `run()` holds a lock when it is processing events. So we could deadlock as we're waiting for the stopGroup to finish.

Reply:
Note: drop StopLock and be okay with the concurrence Stop call undefined behavior.